### PR TITLE
Adding local AUC computation to the evaluation package

### PR DESCRIPTION
**Pull request by XianXing**
As discussed in issue #42, currently the only AUC computation supported in Photon/GAME is ```RDD``` based, e.g., the input scores and labels are in the ```RDD[(Double, Double)]``` format.

However, in order to enable per random effect AUC computation as in issue #41, computing AUC locally in the ```Array[(Double, Double)]``` or ```Seq[(Double, Double)]``` format will be a useful feature to have.

Besides, the local AUC computation will be helpful to our effort in evaluating different matrix factorization packages as well.

One more change made in this PR is that, now the local AUC computation will take the weight of each data point into consideration when computing AUC.

### CODE REVIEW

File: photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/AreaUnderROCCurveLocalEvaluator.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      if (label > MathConst.POSITIVE_RESPONSE_THRESHOLD) {
+        positiveCount += weight
+      } else {
+        aucAccum += positiveCount
+        negativeCount += weight
+      }
+    }
+
+    //normalize ACU over the total area
+    aucAccum / (positiveCount * negativeCount)
```
Comment (joshvfleming) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/AreaUnderROCCurveLocalEvaluator.scala]: @XianXing I believe there's a subtle flaw in this computation. If we have two identical scores with conflicting ground-truth labels, we're accumulating a little too much area for the "true" one. For example, in the following case:

```
labels = [0, 1, 0, 0, 1, 1, 1]
scores = [-0.1, -1, 0, -1, 2, 6, 8]
```

we compute an AUC of **0.8333** here, but it should be **0.7917**.

Comment (joshvfleming) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/AreaUnderROCCurveLocalEvaluator.scala]: @XianXing I believe there's a subtle flaw in this computation. If we have two identical scores with conflicting ground-truth labels, we're accumulating a little too much area for the "true" one. For example, in the following case:

```
labels = [0, 1, 0, 0, 1, 1, 1]
scores = [-0.1, -1, 0, -1, 2, 6, 8]
```

we compute an AUC of **0.8333** here, but it should be **0.7917**.

**Comment by joshvfleming** [comment]
@XianXing I believe there's a subtle flaw in this computation. If we have two identical scores with conflicting ground-truth labels, we're accumulating a little too much area for the "true" one. For example, in the following case:

```
labels = [0, 1, 0, 0, 1, 1, 1]
scores = [-0.1, -1, 0, -1, 2, 6, 8]
```

we compute an AUC of **0.8333** here, but it should be **0.7917**.

**Comment by ashelkovnykov** [comment]
LGTM, aside from the AUC concerns.

**Comment by XianXing** [comment]
Very good point @joshvfleming, the previous algorithm used indeed ignored the case you mentioned. I've updated the algorithm and the unit tests accordingly to cover it. Please take a look.

**Comment by namitk** [comment]
@XianXing: Before merging, could you rebase -i the commits to squash the extraneous, minor fixes? I believe this PR can be two commits, the first one and then the rest merged under "fix bug" or "more rigorous and robust algorithm..."

**Comment by XianXing** [comment]
Good suggestion @namitk, now the commit history is cleaned a bit.

**Comment by joshvfleming** [comment]
Looks good!

**Pull request merged**